File: lyapunov.py
from __future__ import annotations

import numpy as np

from solvers import *


def modify_init_conditions(init_conditions, threshold=1e-6):
    """
    Add a tiny random displacement to each object's position & velocity.
    """
    for obj in init_conditions:
        dp = np.random.uniform(-threshold, threshold, size=3)
        dv = np.random.uniform(-threshold, threshold, size=3)
        obj.position = obj.position + dp
        obj.velocity = obj.velocity + dv
    return init_conditions


class LyapunovCalculator:
    """
    Computes the largest Lyapunov exponent by linear-fitting
    log(separation) vs time.
    """

    def __init__(self, system1: System, system2: System) -> None:
        assert (
            system1.solved and system2.solved
        ), "Both systems must be solved (i.e. integrated) first."
        assert system1.h == system2.h, "Step sizes must match"
        assert system1.num_steps == system2.num_steps, "Number of steps must match"
        self.sys1 = system1
        self.sys2 = system2
        self.h = system1.h
        self.num_steps = system1.num_steps

    def compute_separation(self) -> np.ndarray:
        """
        Returns an array delta of length num_steps+1,
        where delta[i] is the mean distance between the two
        systems' mass configurations at step i.
        """
        # assume system.positions is dict or list: mass_id -> array shape (num_steps+1,3)
        # stack over masses then take mean over masses
        all_dists = []
        for m in self.sys1.positions:
            pos1 = np.asarray(self.sys1.positions[m])  # shape (N+1,3)
            pos2 = np.asarray(self.sys2.positions[m])
            d = np.linalg.norm(pos1 - pos2, axis=1)  # shape (N+1,)
            all_dists.append(d)
        # shape (num_masses, N+1) -> mean over masses
        return np.mean(np.vstack(all_dists), axis=0)

    def calculate_lyapunov_exponent(self) -> float:
        delta = self.compute_separation()
        delta0 = delta[0]
        if delta0 <= 0:
            raise ValueError("Initial separation must be > 0")
        t = np.arange(len(delta)) * self.h
        y = np.log(delta / delta0)
        # np.polyfit returns [slope, intercept]
        lyapunov, intercept = np.polyfit(t, y, 1)
        print("rate of separation:", delta0 * np.e * (lyapunov * len(delta) * self.h))
        return lyapunov

File: src/lyapunov.rs
use crate::mass::Mass;
use crate::solvers::System;
use nalgebra::Vector3;
use rand::Rng;
use rand_distr::Uniform;

pub fn modify_init_conditions(init_conditions: &mut [Mass], threshold: f64) {
    let mut rng = rand::rng();
    let range = Uniform::new_inclusive(-threshold, threshold).unwrap();
    for obj in init_conditions.iter_mut() {
        let pos_perturbation =
            Vector3::new(rng.sample(&range), rng.sample(&range), rng.sample(&range));
        let vel_perturbation =
            Vector3::new(rng.sample(&range), rng.sample(&range), rng.sample(&range));
        obj.position += pos_perturbation;
        obj.velocity += vel_perturbation;
    }
}

pub struct LyapunovCalculator {
    system1: System,
    system2: System,
}

impl LyapunovCalculator {
    pub fn new(system1: System, system2: System) -> Self {
        assert_eq!(system1.h, system2.h, "Step sizes must match");
        assert_eq!(
            system1.num_steps, system2.num_steps,
            "Number of steps must match"
        );
        LyapunovCalculator { system1, system2 }
    }

    pub fn compute_separation(&self) -> Vec<f64> {
        let num_steps = self.system1.positions[0].len();
        let num_masses = self.system1.objs.len();
        let mut separations = vec![0.0; num_steps];

        for step in 0..num_steps {
            let mut total_sep = 0.0;
            for mass in 0..num_masses {
                let pos1 = self.system1.positions[mass][step];
                let pos2 = self.system2.positions[mass][step];
                let sep = (pos1 - pos2).norm();
                total_sep += sep;
            }
            separations[step] = total_sep / num_masses as f64;
        }
        separations
    }

    pub fn calculate_lyapunov_exponent(&self) -> f64 {
        let separations = self.compute_separation();
        let initial_sep = separations[0];
        assert!(
            initial_sep > 0.0,
            "Initial separation must be greater than zero"
        );

        let num_steps = separations.len();
        let total_time = num_steps as f64 * self.system1.h;
        let sum: f64 = separations
            .iter()
            .skip(1)
            .map(|&sep| (sep / initial_sep).ln())
            .sum();
        sum / total_time / (num_steps - 1) as f64
    }
}
